### DISCUSSION TITLE
How to query a single item by field content when id is unknown ?

### DISCUSSION THREAD

**Discussion by martinval** [directus/directus#2718] [Archive]
I have a list of items where I want to get that has a unique field content (eg. name or email), because I don't know the id.

For the API, the docs show this way: GET /:project/items/:collection/:id
Is it possible to get a record in another way ? with something like:  
/:project/items/:collection/?email="test@test.com" ?

The same applies to GraphQL, how to pass variables ?

In PHP, I've tried something like, but I get an error:
```
$email = "test@test.com";
$query =
      query ($emailVar: $email){
            users (filter: {email_eq: $emailVar} ) {
              data {
                id
                firstname
                lastname
                ....
              }
            }
        }

```
Thanks in advance for any pointers.

**Comment by rijkvanzanten** [comment]
You can use a filter in combination with the `single` param:

```
GET /items/:collection?filter[email]=test@test.com&single=1
```

**Reply by martinval** [reply]
Thanks @rijkvanzanten, I dug further into the docs and found the filters eventually :)

**Reply by trojancode** [reply]
it's not working now

**Reply by rijkvanzanten** [reply]
@trojancode there's no single param in v9, just use `limit=1` 👍🏻

**Reply by trojancode** [reply]
but it returns as array

**Reply by simonmaris** [reply]
jap, Is there any way not to get an array returned? (graphQL here, but same behaviour). It's annoying to deal with the array in the frontend code while we know it's just one element. Any ideas? Thanks!

**Reply by rijkvanzanten** [reply]
No, if you search/filter, you get the results as an array

**Reply by simonmaris** [reply]
Sure, but with a limit=1 couldn't the array be flattened?

**Reply by rijkvanzanten** [reply]
Not really, as that changes the type signature, which causes issues in things like TypeScript, OpenAPI, and even GraphQL, as that type of "item | item[]" output type isn't allowed. In order to do that, we'd have to add another suffixed field into graphql, so you end up with something like:

```graphql
query {
  articles(filter: { status: { _eq: 'published' }}) { # multiple
    id
  }

  articles_single(filter: { status: { _eq: 'published' }}) { # as a single object
    id
  }
}
```

which isn't great in it's own right 🤔

**Reply by simonmaris** [reply]
Thanks for the in-depth reply, very much appreciated! I was pretty unaware of those complications..

**Reply by rijkvanzanten** [reply]
Yeah.. graphql can be surprisingly strict in things like that

**Reply by TheRealMaN** [reply]
Does this works now "GET /items/:collection?filter[email]=test@test.com&limit=1"? I get error "\"value\" must contain at least one of [query, keys]"

**Reply by rijkvanzanten** [reply]
@TheRealMaN are you sure you're doing a GET request?

**Reply by TheRealMaN** [reply]
I needed to update an item by the certain field, not knowing the id, what I come up with:
PATCH https://directus.app/items/:collection/?filter[cardNumber][_eq]=value&limit=1
payload
```
{
    "query": {
        "cardNumber": "value"
    },
    "data": {
        "date": "2022-12-29",
        "balance": 6000
    }
}
```
